src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileLibraryType.cs

```diff
@@ -11,6 +11,22 @@ namespace NuGet.VisualStudio.SolutionExplorer.Models
     internal enum AssetsFileLibraryType : byte
     {
         Package,
-        Project
+
+        Project,
+
+        /// <summary>
+        /// When it is unknown whether the library represents a package or a project.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// This occurs, for example, when a referenced item is not found. It will be present in the lock file's
+        /// messages, but not elsewhere in the lock file.
+        /// </para>
+        /// <para>
+        /// This enum member exists only to support attaching diagnostic items to their package or project reference
+        /// node in the tree.
+        /// </para>
+        /// </remarks>
+        Unknown
```

Comment:
I think using `Unknown` can be confusing because the non-existant type is either `Package` or `Project`. To make the code easier to understand, we could consider adding a readonly property called `LibraryType` of type `AssetsFileLibraryType` to `AssetsFileLogMessage`. This would allow us to use `AssetsFileLogMessage.LibraryType` instead of `Unknown` when creating a placeholder or dummy `AssetFileTargetLibrary` object. By implementing this change, we could eliminate the need to check for `Unknown` in `TryGetPackage` or `TryGetProject` methods.

Reply:
Unfortunately we still wouldn't know what value to put in a `AssetsFileLogMessage.LibraryType` property. The `project.assets.json` file doesn't contain enough information to model that. Hence the `Unknown` value.

For example, here's are the log messages for a package and project, respectively:

```json
"logs": [
  {
    "code": "NU1101",
    "level": "Error",
    "message": "Unable to find package NonExistantPackage. No packages exist with this id in source(s): C:\\Program Files\\dotnet\\library-packs, Microsoft Visual Studio Offline Packages, NuGet local, nuget.org",
    "libraryId": "NonExistantPackage",
    "targetGraphs": [
      "net7.0"
    ]
  },
  {
    "code": "NU1104",
    "level": "Error",
    "message": "Unable to find project 'D:\\repos\\issues\\Issue8947UnresolvedPackageWarningMissing\\NonExistantProject\\NonExistantProject.csproj'. Check that the project reference is valid and that the project file exists.",
    "libraryId": "D:\\repos\\issues\\Issue8947UnresolvedPackageWarningMissing\\NonExistantProject\\NonExistantProject.csproj",
    "targetGraphs": [
      "net7.0"
    ]
  }
]
```

Reply:
Sorry for the confusion. I should have been more clearer. Given that `Path.IsRooted("libraryId")` will be `true` for projects but not for packages.   

```cs
    /// <summary>
    /// Models a diagnostic found in the assets file. Immutable.
    /// </summary>
    internal readonly struct AssetsFileLogMessage
    {
        public AssetsFileLogMessage(string projectFilePath, IAssetsLogMessage logMessage)
        {
            Code = logMessage.Code;
            Level = logMessage.Level;
            WarningLevel = logMessage.WarningLevel;
            Message = logMessage.Message;
            LibraryName = NormalizeLibraryName(logMessage.LibraryId, projectFilePath);
            LibraryType = Path.IsRooted(logMessage.LibraryId) ? AssetFileTargetLibrary.Project : AssetFileTargetLibrary.Package
        }
    //remaining code
    }
```

With these changes, when we are creating the dummy/placeholder object, we can pass logmessage.LibraryType instead of using Unknown as the `LibraryType` property of `AssetFileTargetLibrary`

Reply:
Oh I see. I guess it comes down to whether you think and absolute path is a good indicator that is a project or package. I don't know it to be, so prefer to capture that uncertainty in the data. It also means the unknown flag helps explain why so many other fields in the object have no value.

From a design point of view I prefer modelling that uncertainty explicitly, even if it means an extra check in two places.

A better fix overall would be for the lock file to include the library type in the log message JSON. That must be known by the writer of the file, but then that information is not serialised and is therefore lost to consumers.